### ISSUE TITLE
[feature request] `proton radiography diagnostic` with an additional `grid` for externally applied magnetic field

### ISSUE THREAD

**Issue by weipengyao** [PlasmaPy/PlasmaPy#1725]
Dear developers,

Now that more and more research groups are coupling their laser-target interaction with externally applied stable (~us), large volume (~cm^3), and strong (~10s T) magnetic field, taking account of this external B-field in the proton radiography (PR) diagnostic seems more and more crucial for the proper interpretation (for example, a recent paper: https://aip.scitation.org/doi/10.1063/5.0054172)

The PR tool in `PlasmaPy` has a `grid` between the `source` and the `detector`, in which we can add the EM fields, generated from the laser-target interaction, etc. It typically has a volume of mm^3, with a resolution of 10s micron.

Normally, we apply the simulated or analytical EM fields into this `grid` and check the results from the detector. 
However, with the applied magnetic field, we need one more grid with cm^3 in volume and, probably, lower resolution, to allow the source particles (i.e., p+) to propagate through, before they reach the detector.

Is it possible to add such a feature in the PR tool, that allows the consideration of the externally applied magnetic fields?

Thanks very much!
Weipeng YAO

**Comment by pheuer** [comment]
Hello! 

Yes, I strongly agree that this would be a useful addition, and I've been playing with how to do it for a while. I think the best solution is to allow multiple grids to be provided in an ordered list. If particles are on the first grid, then the fields should be interpolated from that grid. If not, the code should then check if they are on the next grid in the list, etc. So, for your example, the small micron scale grid would be first, followed by the cm-scale background field grid. 

I can bump this up my queue to work on, but if you or a student would like to try to make the modifications as a new pull request I would be very happy to work with you on that!

**Comment by weipengyao** [comment]
Hi Peter, 

Thanks for your reply and support. I would be happy to give it a try with your guidance.

Just one comment about my example case -- since the micron-scale grid is immersed inside the coil that generates the B-field, so even before the small grid, there is already some B-field (with not-so-small gradient).
For example, a typical spatial distribution of the B-field can be found in this paper, Fig. 5 (https://aip.scitation.org/doi/10.1063/1.4795551)

<img width="867" alt="image" src="https://user-images.githubusercontent.com/37392893/188655908-e46d2d59-e471-4692-9fee-6f4f04df7158.png">


But I guess the essential strategy is still the same. 

So, where should I start with?

Thanks again!

**Comment by pheuer** [comment]
Yes, so for this to work you would have to blend the two fields together so that there is not a discontinuity. Another option would be to interpolate the field for each grid that the particle is on and then add them together. Then you would basically subtract the background field from the high resolution grid and apply a mask at the edges to blend them together. The effect should be that the field from each particle gets interpolated based on which grid it is on. The downside is that many particles will then be interpolated onto TWO grids, which will be slower. 

Does that make sense? 

First things first, you should [install PlasmaPy for development](https://docs.plasmapy.org/en/stable/contributing/install_dev.html). Depending on your familiarity with git, etc. you may want to watch some of the [tutorials on the PlasmaPy youtube channel](https://www.youtube.com/watch?v=sv6Raq_aI8U&t=2846s). 

The actual code for this function is all in this file: https://github.com/PlasmaPy/PlasmaPy/blob/main/plasmapy/diagnostics/charged_particle_radiography.py

As far as the code goes, the following modifications are needed (as a start): 

- Allow a list of 'grid' objects to be passed into the 'grid' argument. If only one grid is passed, wrap it in a list so that there is always a list. E.g. 

```python
if isinstance(grid, AbstractGrid):
    grid = [grid, ]
```

- Modify this line to calculate the max theta hit grid for each grid in the list, then take the maximum

```python
self.max_theta_hit_grid = self._max_theta_hit_grid()
```

Similarly, for the rest of the `__init__ method`, modify so that all the validation etc. for the grids is being run on each grid in the list. 


- Modify `_coast_to_grid` to calculate `dist` to each grid and then coast to the minimum distance (e.g. the edge of the largest grid)

- Within `_push`, run `on_grid` on each of the grids. `self.on_grid` will need to be a list of lists now, recording which particles are on each grid. `self.entered_grid` should probably also be a list of lists to keep track of which particles entered which of the grids. 

- (The big one): the interpolation section of  `_push` needs to be modified so that you interpolate the correct fields for the correct particles and then sum those fields as I proposed above. This will involve looping through the list of grids and calling the interpolators on the appropriate subsets of the particles (as determined above by the `on_grid` method. 

- Modify `_adaptive_dt` to calculate the grid resolution of each grid _that currently has particles on it_  based on the ongrid method. We'll need a different dt for each grid. The timestep for each particle needs to be governed by the highest resolution grid it is on. E.g. if it is on both, the smaller dt wins. 

- Modify `_push` to push the particles in batches based on which timestep they have. If we push everything based on the smaller timestep, the particles will take a very long time to move through the large length scale low res grid. So, sort out the particles based on timestep and then push them separately. 

- Modify `_stop_condition` as necessary to deal with changes to `entered_grid`. We want to know what fraction of the particle have entered A grid and are now on NO grid. 

- Implement at least one test with two grids of different scales and an expected analytical outcome and demonstrate that the new functionality works. That will go in this file: https://github.com/PlasmaPy/PlasmaPy/blob/main/plasmapy/diagnostics/tests/test_charged_particle_radiography.py

You should create a fork of PlasmaPy, create a branch for this project, and then open a pull request (marked 'draft'). That will allow me to see your code and provide line-by-line comments and feedback. Eventually, once it is ready, that pull request can then be merged into the main PlasmaPy branch to make the functionality available!

**Comment by weipengyao** [comment]
Dear Peter, 
Thanks for this very detailed response. 
I will keep you updated on my progress here. 
Cheers

**Comment by weipengyao** [comment]
Dear @pheuer, 

I tried to follow your guidance one-by-one, and opened a PR #1730. 
(I thought it would have the same id as this issue, e.g. 1725?)

Anyway, at the very beginning, changing the `grid` into list, I am not sure how to do that -- Just by adding your suggestions:
```
if isinstance(grid, AbstractGrid):
    grid = [grid, ]
```
before the `self.grid = grid`? 

Could you show me a few examples so that I can follow them and modify the rest of the `__init__`. 

Thanks.

**Comment by pheuer** [comment]
Hello, 

No worries - the issue and PR will have different IDs (since different people could open PRs for the same issue) but github automatically connects the two (see above where a link has been added to your PR). 

The idea is to make `self.grid` always a list (even if only one grid is given) so that for the rest of the code you can always treat it like a list. So, for example: 


```python

if isinstance(grid, AbstractGrid):
    self.grid = [grid, ]

elif isinstance(grid, list):
    self.grid = grid

```


This code will handle two situations: 

1) If the user provides a single grid, we get a list out with one grid in it
2) If the user provides a list of grids, we get a list with multiple grids in it. 

Does that make sense?

**Comment by weipengyao** [comment]
Ah, yes, it is clear to me now. 

Then, for example the next line of code `self.grid_arr = grid.grid.to(u.m).value`

Now that `grid` becomes a list, what is the appropriate way to change it accordingly?

Thanks!

**Comment by pheuer** [comment]
So, now every attribute referring to grid will have to be a list too. So, you will have to add loops like 

```python
self.grid_arr = []
for grid in self.grid:
    self.grid_arr.append(grid.grid.to(u.um).value)
```

It may be worth adding a plural to make this clearer to future readers, e.g. `self.grid` -> `self.grids` and so on for all these variables. That will also make sure you don't miss any!

**Comment by weipengyao** [comment]
Ah! Yes, thanks very much. Now I know how to proceed! 
And that is indeed a good trick, i.e., adding plurals. 

I will keep converting those needed into lists and keep you updated!

**Comment by weipengyao** [comment]
Hi @pheuer 

I am on this task now:

>Within _push, run on_grid on each of the grids. self.on_grid will need to be a list of lists now, recording which particles are on each grid. self.entered_grid should probably also be a list of lists to keep track of which particles entered which of the grids.

How should I do with this line to make it a list of lists?
```python3
self.on_grid = self.grid.on_grid(pos)
```

For example, something like this below?

```python3
self.on_grids = [ ]
for grids in self.grids:
    self.on_grids.append(grids.on_grid(pos))
```
Thanks!

**Comment by pheuer** [comment]
Yes, that will work! You could also use list comprehensions to make this even neater (I probably should have suggested that above for `grid_arr` too):

```python
self.on_grids = [ grid.on_grid(self.pos) for grid in self.grids ]
```
